[Security Solution][Explore][Overview Dashboard] Move from viz responses to viz tables

kapral18 opened the pull request:
Addresses #214733

This PR revamps the visualization response handling.  In most of the affected files `useVisualizationResponse` was purely used to obtain total count of docs to just verify if data exists. But to support that we had to juggle with response type which is not easy to determine and is too blanket. Unfortunately the type being too generic led to a couple of issues described in the above mentioned ticket.

I initially tried to improve the types but the more I digged the more I realized that it's very hard to do because it's not us who makes the request but the Lens and request and response is actually an implementation detail of Lens component. We control the request indirectly by providing [lens attributes](https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/security_solution/public/common/components/visualization_actions/lens_attributes/common/events.ts#L17). And they define the request and affect the expected shape of response. But since exact logic of determining the shape is hidden in Lens we cannot rely on it and cannot narrow it down with generics on our side.

So turns out Lens also exposes another object `TablesAdapter.tables` that contains tabular data and meta data about the request and response. From there we can access the total count of each lens response. And its structure is more  mappable to lens attributes structure. Though it has its own nuancing and legacy deviations.

Still this is a more idiomatic way of accessing lens total counts atm.

### Fixes v9+

## Before

https://github.com/user-attachments/assets/cf1c8d06-dd37-4cd9-9718-3ecb6347e342

## After

https://github.com/user-attachments/assets/99de06dc-f04c-40da-bd0c-3910b091146c

### Fixes v8.18+

## Before with bsearch

https://github.com/user-attachments/assets/4a4d3a0c-5c80-456a-ad8e-d0edf7f2a665

## After

https://github.com/user-attachments/assets/bf3530b0-6fb3-44d9-a159-eebe68b1fa9e



<!--ONMERGE {"backportTargets":["8.17","8.18","8.19","9.0"]} ONMERGE-->

kapral18 committed:
[Security Solution][Explore][Overview Dashboard] Handle undefined and empty responses in alert histogram count

Addresses #214733

- Modify `useAlertHistogramCount` to use optional chaining and nullish coalescing for safer access to response data
- Update test cases to cover scenarios with undefined, empty array, and incomplete response objects

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-threat-hunting-explore (Team:Threat Hunting:Explore)

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

semd reviewed (approved):
LGTM

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/285604)
* Commit: e92e227c190ee926319c33102a9f8b7412ff25b2

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/64743b3a820e3af4478d78bb643fa3531b302aa6) | [after](https://github.com/elastic/kibana/commit/e92e227c190ee926319c33102a9f8b7412ff25b2) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 8.9MB | 8.9MB | +94.0B |



cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/285604","number":285604,"commit":"e92e227c190ee926319c33102a9f8b7412ff25b2"}],"number":285604}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/286351)
* Commit: 40fb043f49717335970ee98c8e5270dd00f1f28e

### Failed CI Steps
* [Serverless Investigations - Security Solution Cypress Tests #3](https://buildkite.com/elastic/kibana-pull-request/builds/286351#0195c06e-ef08-4a28-882f-f33ae804cdbd)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/286351#0195c06e-ef08-4a28-882f-f33ae804cdbd) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/286351/jobs/0195c06e-ef08-4a28-882f-f33ae804cdbd/artifacts/0195c0a7-7777-4265-a793-f0123ba521c0) Serverless Investigations - Security Solution Cypress Tests #<span></span>3 / KPI visualizations in Alerts Page Histogram legend hover actions should add a filter out to KQL bar should add a filter out to KQL bar
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/286351#0195c06e-ef08-4a28-882f-f33ae804cdbd) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/286351/jobs/0195c06e-ef08-4a28-882f-f33ae804cdbd/artifacts/0195c0a7-7777-4e0b-8f3d-8ffc53fdfa18) Serverless Investigations - Security Solution Cypress Tests #<span></span>3 / KPI visualizations in Alerts Page Histogram legend hover actions should add To Timeline should add To Timeline

### History
* :green_heart: [Build #285604](https://buildkite.com/elastic/kibana-pull-request/builds/285604) succeeded e92e227c190ee926319c33102a9f8b7412ff25b2

cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/286351","number":286351,"commit":"40fb043f49717335970ee98c8e5270dd00f1f28e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/285604","number":285604,"commit":"e92e227c190ee926319c33102a9f8b7412ff25b2"}],"number":286351}
buildkite-pr-comment-->

kapral18 committed:
chore: move from responses to tables and cleanup

- Removed unused `setQuery` prop from various components including
`EventsQueryTabBodyComponent`, `MatrixHistogramComponent`,
`StatefulTopNComponent`, and others.
- Updated tests to reflect the removal of `setQuery` prop.
- Introduced `getTotalCountFromTables` utility to calculate total count
from all tables layers.
- Updated useVisualizationResponse to return `tables` for data handling.
- Adjusted subtitle logic in components to use new data handling
approach.
- Removed unused mock data and updated existing tests to align with the
new implementation.

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/297036)
* Commit: 5cb3ee574764d4217406805b812e48aac6ec6dba

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/297036#019680e3-b880-4d7d-91e0-998e3b9aabf1)

### History
* :broken_heart: [Build #286351](https://buildkite.com/elastic/kibana-pull-request/builds/286351) failed 40fb043f49717335970ee98c8e5270dd00f1f28e
* :green_heart: [Build #285604](https://buildkite.com/elastic/kibana-pull-request/builds/285604) succeeded e92e227c190ee926319c33102a9f8b7412ff25b2

cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297036","number":297036,"commit":"5cb3ee574764d4217406805b812e48aac6ec6dba"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/286351","number":286351,"commit":"40fb043f49717335970ee98c8e5270dd00f1f28e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/285604","number":285604,"commit":"e92e227c190ee926319c33102a9f8b7412ff25b2"}],"number":297036}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/297056)
* Commit: a86167d2ef99ae6868b0cb36d0e240432ab7e9a7

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/297056#0196810e-46a8-4f61-9b1b-567e2f04989e)

### History
* :broken_heart: [Build #297036](https://buildkite.com/elastic/kibana-pull-request/builds/297036) failed 5cb3ee574764d4217406805b812e48aac6ec6dba
* :broken_heart: [Build #286351](https://buildkite.com/elastic/kibana-pull-request/builds/286351) failed 40fb043f49717335970ee98c8e5270dd00f1f28e
* :green_heart: [Build #285604](https://buildkite.com/elastic/kibana-pull-request/builds/285604) succeeded e92e227c190ee926319c33102a9f8b7412ff25b2

cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297056","number":297056,"commit":"a86167d2ef99ae6868b0cb36d0e240432ab7e9a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297036","number":297036,"commit":"5cb3ee574764d4217406805b812e48aac6ec6dba"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/286351","number":286351,"commit":"40fb043f49717335970ee98c8e5270dd00f1f28e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/285604","number":285604,"commit":"e92e227c190ee926319c33102a9f8b7412ff25b2"}],"number":297056}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/297083)
* Commit: fd0bbdf258ba12646497ba3e3c6dfac0ca9d7855

### Failed CI Steps
* [Serverless Detection Engine - Exceptions - Security Solution Cypress Tests #4](https://buildkite.com/elastic/kibana-pull-request/builds/297083#01968160-f0cf-4a71-b4e8-b6ab3d7da827)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/297083#01968160-f0cf-4a71-b4e8-b6ab3d7da827) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/297083/jobs/01968160-f0cf-4a71-b4e8-b6ab3d7da827/artifacts/01968193-058d-4175-b21c-968a1a383b34) Serverless Detection Engine - Exceptions - Security Solution Cypress Tests #<span></span>4 / Close matching Alerts  Should close all alerts from if several rules has shared exception list Should close all alerts from if several rules has shared exception list

### History
* :broken_heart: [Build #297056](https://buildkite.com/elastic/kibana-pull-request/builds/297056) failed a86167d2ef99ae6868b0cb36d0e240432ab7e9a7
* :broken_heart: [Build #297036](https://buildkite.com/elastic/kibana-pull-request/builds/297036) failed 5cb3ee574764d4217406805b812e48aac6ec6dba
* :broken_heart: [Build #286351](https://buildkite.com/elastic/kibana-pull-request/builds/286351) failed 40fb043f49717335970ee98c8e5270dd00f1f28e
* :green_heart: [Build #285604](https://buildkite.com/elastic/kibana-pull-request/builds/285604) succeeded e92e227c190ee926319c33102a9f8b7412ff25b2

cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297083","number":297083,"commit":"fd0bbdf258ba12646497ba3e3c6dfac0ca9d7855"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297056","number":297056,"commit":"a86167d2ef99ae6868b0cb36d0e240432ab7e9a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297036","number":297036,"commit":"5cb3ee574764d4217406805b812e48aac6ec6dba"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/286351","number":286351,"commit":"40fb043f49717335970ee98c8e5270dd00f1f28e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/285604","number":285604,"commit":"e92e227c190ee926319c33102a9f8b7412ff25b2"}],"number":297083}
buildkite-pr-comment-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

kapral18 committed:
chore: return loading because of error state

kapral18 committed:
chore: properly handle error case for tables.tables being empty object

- Add a new test file `use_embeddable_inspect.test.tsx` to cover the `useEmbeddableInspect` hook
- Test scenarios include:
  - Calling `onEmbeddableLoad` with adapter data when loaded
  - Handling loading state with empty requests/responses
  - Ensuring no tables are included when `adapters.tables.tables` is empty
  - Verifying no errors occur when `onEmbeddableLoad` is not provided
- Update `use_embeddable_inspect.tsx` to improve handling of loading state and empty tables
- Ensure `onEmbeddableLoad` is called with correct data structure and conditions

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/297385)
* Commit: 4144d8247d39cce3ae25c62be5833b4ace094d73

### Failed CI Steps
* [Serverless Detection Engine - Exceptions - Security Solution Cypress Tests #4](https://buildkite.com/elastic/kibana-pull-request/builds/297385#019683a7-ebe0-4fee-8803-e04fa5cee7c3)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/297385#019683a7-ebe0-4fee-8803-e04fa5cee7c3) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/297385/jobs/019683a7-ebe0-4fee-8803-e04fa5cee7c3/artifacts/019683d7-34cb-470e-973c-bb5f0b586f6c) Serverless Detection Engine - Exceptions - Security Solution Cypress Tests #<span></span>4 / Close matching Alerts  Should close all alerts from if several rules has shared exception list Should close all alerts from if several rules has shared exception list

### History
* :broken_heart: [Build #297083](https://buildkite.com/elastic/kibana-pull-request/builds/297083) failed fd0bbdf258ba12646497ba3e3c6dfac0ca9d7855
* :broken_heart: [Build #297056](https://buildkite.com/elastic/kibana-pull-request/builds/297056) failed a86167d2ef99ae6868b0cb36d0e240432ab7e9a7
* :broken_heart: [Build #297036](https://buildkite.com/elastic/kibana-pull-request/builds/297036) failed 5cb3ee574764d4217406805b812e48aac6ec6dba
* :broken_heart: [Build #286351](https://buildkite.com/elastic/kibana-pull-request/builds/286351) failed 40fb043f49717335970ee98c8e5270dd00f1f28e
* :green_heart: [Build #285604](https://buildkite.com/elastic/kibana-pull-request/builds/285604) succeeded e92e227c190ee926319c33102a9f8b7412ff25b2

cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297385","number":297385,"commit":"4144d8247d39cce3ae25c62be5833b4ace094d73"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297083","number":297083,"commit":"fd0bbdf258ba12646497ba3e3c6dfac0ca9d7855"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297056","number":297056,"commit":"a86167d2ef99ae6868b0cb36d0e240432ab7e9a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297036","number":297036,"commit":"5cb3ee574764d4217406805b812e48aac6ec6dba"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/286351","number":286351,"commit":"40fb043f49717335970ee98c8e5270dd00f1f28e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/285604","number":285604,"commit":"e92e227c190ee926319c33102a9f8b7412ff25b2"}],"number":297385}
buildkite-pr-comment-->

kapral18 committed:
chore: fix cypress regressions

- initially I replaced lens-embeddable logic for detecting no response
  with a working implementation. But that only uncovered the fact that
  it was seemingly misput in the first place. It never actually hit the
  responses != null && responses.length === 0 condition because
  responses was never empty, even if the response was empty so it was a
  dead branch. When I fixed it I actually enabled this path and it
  broken cypress tests as it should've. So I removed the if check
  altogether for empty state so that empty state can be rendered by lens
  as it does now in all UIs.

- removed the respective dead test

- also normalized the inspect assignment logic during use
  onEmbeddableLoad invocation, so that if request is loading
  we set the empty state of inspect as expected

rylnd commented on x-pack/solutions/security/plugins/security_solution/public/detection_engine/rule_creation_ui/components/rule_preview/preview_histogram.test.tsx:
```diff
@@ -40,14 +41,34 @@ jest.mock('../../../../common/utils/normalize_time_range');
+  ),
+  useVisualizationResponse: jest
+    .fn<ReturnType<typeof useVisualizationResponse>, []>()
+    .mockReturnValue({
```
Since the API for this hook is expanding, maybe now is the time to provide an accompanying mock response for the hook? "Mock with a well-formatted response" is really the only thing this test cares about (and only now that we're passing the response to `getTotalCountFromTables`, here).

kapral18 replied:
Hey @rylnd thanks for checking.

> Since the API for this hook is expanding

Can you elaborate? 

If you mean response api surface is expanding it's actually the opposite it's shrinking because we removed `requests` and `responses` properties  and added `tables`.

> maybe now is the time to provide an accompanying mock response for the hook

Can you clarify wdym exactly and why?

rylnd replied:
It's a common pattern to provide a mock object along with the exported code; either a fully mocked object, mocked return values, or both. [Here](https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/security_solution/public/common/hooks/use_app_toasts.mock.ts) is an example for a hook. The idea is to remove all the implementation details from this test, and replace it with something like:

```ts
import useVisualizationResponseMock from './use_visualization_response.mock';

const useMock = useVisualizationResponseMock.create(); // will have a default `mockReturnValue`
useMock.mockReturnValue(useVisualizationResponseMock.buildEmptyReturnValue()); // mock named by use case
```

The advantages here are that a) the test consumer doesn't need to know/care about the implementation of the hook itself in order to mock it for their test, and b) the test likely wouldn't need to change in the future unless the semantics of the hook change (i.e. if "empty return value" is no longer makes sense).

kapral18 replied:
Thanks for clarifying. I like the approach. I've been mostly using response factories like createFooBar(...opts) in similar accompanying files before elastic, but I didn't notice similar stuff being used here, so good to know.

Is there a commonly agreed DSL for these method names? or is it free form?

rylnd replied:
If you search security for `import.*@kbn.*mock` you'll find some good examples.

`import { coreMock } from '@kbn/core/public/mocks';` is probably the de facto for a plugin-level mock, but it's a little looser for individual services.

rylnd commented on x-pack/solutions/security/plugins/security_solution/public/detection_engine/rule_creation_ui/components/rule_preview/preview_histogram.tsx:
```diff
@@ -97,11 +98,11 @@ const PreviewHistogramComponent = ({
   const previewQueryId = `${ID}-${previewId}`;
   const previewEmbeddableId = `${previewQueryId}-embeddable`;
-  const { responses: visualizationResponses } = useVisualizationResponse({
+  const { tables } = useVisualizationResponse({
     visualizationId: previewEmbeddableId,
   });
```
Since this seems to be a core use case for the hook (as you stated in the PR description), did you consider adding `totalCount` as part of the hook's response, instead of making it external? Then the hook would not have to expose the table abstraction completely for these "total" consumers:

```suggestion
  const { getTotalHits } = useVisualizationResponse({
    visualizationId: previewEmbeddableId,
  });
  
  const totalCount = getTotalHits();
```

kapral18 replied:
hey @rylnd that's a great point.

I did indeed consider it initially. But I am actually assuming here that folks are ok with summing all lens layer totalCounts. For example you can have 2 separate chart renderings with separate labels that you want to control independently. And for that you can access `tables[layerId].meta.statistics.totalCount` directly if you have a layerId at hand for xy charts which you do because it's in the lens attributes file. Or create your own extraction logic.

And since I wasn't sure what is the better way and didn't want to decide the use case for all teams I decided to extract that logic out right next and let others choose to use it if it makes sense for their use case or choose to access table directly or with a different parser. 

Also with exposure of `tables` I wanted to emphasize the fact that we now work with datatable structure instead of regular req and response objects. They are very different in their representations and datables have nuances.

And on a minor note I would stay away from naming it getTotalHits() because it uses the language of non datatable response aka `response.hits.total` shape which can be confusing because the expected shapes don't match.  `response.hits.total` can be both a number and an object while totalCounts is always a number. And I know we can fix it by enforcing the types, but that's not the point. 

Domain language is the point and to me `getTotalHits` seems to be associated with regular request response shape. I also saw something similar used in your code with `getTotalHitsValue` which kind of corroborates the point.

Basically I would want to avoid confusing folks that might think this is a regular response shape's total hits so if their next move is to access more details from response they could start searching for entrypoint to aggregations and checking buckets in the response itself and will instead find tables which is breaking the expectation. To me it's always good to be as expressive and unambiguous in your api description as possible wdyt?

And finally we could just incorporate totalCounts into tables response by introducing a container of tables like 

```json
tables: {
  tables: {  [layer1]: {...}}
  meta: { statistics: { totalCounts: 999}}
}
```

To just mirror the structure of each individual layer meta totalcounts and get rid of the external helper.  To me this is a little more idiomatic

Let me know if my thoughts make sense to you.

kapral18 replied:
hey @rylnd that's a great point.

I did indeed consider it initially. But I am actually assuming here that folks are ok with summing all lens layer totalCounts. For example you can have 2 separate chart renderings with separate labels that you want to control independently. And for that you can access `tables[layerId].meta.statistics.totalCount` directly if you have a layerId at hand for xy charts which you do because it's in the lens attributes file. Or create your own extraction logic.

And since I wasn't sure what is the better way and didn't want to decide the use case for all teams I decided to extract that logic out right next and let others choose to use it if it makes sense for their use case or choose to access table directly or with a different parser. 

Also with exposure of `tables` I wanted to emphasize the fact that we now work with datatable structure instead of regular req and response objects. They are very different in their representations and datables have nuances.

And on a minor note I would stay away from naming it getTotalHits() because it uses the language of non datatable response aka `response.hits.total` shape which can be confusing because the expected shapes don't match.  `response.hits.total` can be both a number and an object while totalCounts is always a number. And I know we can fix it by enforcing the types, but that's not the point. 

Domain language is the point and to me `getTotalHits` seems to be associated with regular request response shape. I also saw something similar used in your team code with `getTotalHitsValue` which kind of corroborates the point.

Basically I would want to avoid confusing folks that might think this is a regular response shape's total hits so if their next move is to access more details from response they could start searching for entrypoint to aggregations and checking buckets in the response itself and will instead find tables which is breaking the expectation. To me it's always good to be as expressive and unambiguous in your api description as possible wdyt?

And finally if we still decide to internalize it we could just incorporate totalCounts into tables response by introducing a wrapper for tables like 

```ts
{ 
  tables: {
    tables: {  [layer1]: {...}}
    meta: { statistics: { totalCounts: 999}}
  }
}
```

To just mirror the structure of each individual layer meta totalCounts and get rid of the external helper.  To me this is a little more idiomatic. Let me know wdyt?

rylnd replied:
Those are all fair points! I was mostly trying to minimize the changes needed here, particularly to the tests (which were closely tied to the implementation far before this PR). I don't disagree with returning `tables`, nor about exporting the helper method for more "manual" purposes.

Mainly I was just trying to make the "I only care about the total count" use case simpler and easier, since it seems to be a common one. Having to import and invoke an extra function isn't terrible, but as you've seen it makes more work (and more implementation details) in the tests, as well.

IMO the cleanliness of moving that logic into the hook is well worth the cost; as long as the total calculation is deferred, the only real "cost" is memory for the function definition: consumers are free to call or ignore it as needed. I suspect this is where we might disagree, which is fine.

P.S. I admit that `getTotalHits` was not a great example name, especially since `getTotalCountFromTables` is already established 😉.

kapral18 replied:
I absolutely appreciate your points as well.

> but as you've seen it makes more work (and more implementation details) in the tests, as well.
  
I agree with this.

> the cleanliness of moving that logic into the hook is well worth the cost

I agree as well. Not opposing at all. Was just highlighting how I came to that conclusion in those circumstances.

>  the only real "cost" is memory for the function definition

As you correctly inferred I do disagree but again because imho miniscule memory impact does not justify adding an arguably preventable interface variation. We already have totals as numbers, so having a total of totals as a number is a more natural expectation, especially given that the calculation itself is going to be wrapped in useMemo and computational impact will be reduced further. 

And lastly Imho we shouldn't be even debating this, this total totalcount should come out of the box as part of tablesAdapter. But since it's not I think what I suggested should work. I will try and if it doesn't I will use your suggestion.

Thanks

kapral18 replied:
Done

rylnd reviewed (commented):
I had a suggestion for improving the interface of the `useVisualizationResponse` hook; let me know what you think!

angorayc commented:
Hi Karen, nice enhancement, looking great overall, thank you.
I found the behaviour of topN in the `Detection alerts` was changed in this PR. The data source included more than just alert's index in this PR. This rendered the same chart between `Raw events` and `Detection alerts`. Please double check.


Previous behaviour:
<img width="2556" alt="Screenshot 2025-05-01 at 12 26 13" src="https://github.com/user-attachments/assets/b8fe5997-324d-41cb-9882-b4a5ab05bd26" />

https://github.com/user-attachments/assets/fe05b83d-d00c-42ed-8994-c97b0c4de924


In this PR:

https://github.com/user-attachments/assets/0d160690-108f-4ac7-bd71-f39c5fe5b4d9

angorayc reviewed (commented):
Hi Karen, nice enhancement, looking great overall, thank you.
I found the behaviour of topN in the `Detection alerts` was changed in this PR. The data source included more than just alert's index in this PR. This rendered the same chart between `Raw events` and `Detection alerts`. Please double check.


Previous behaviour:
<img width="2556" alt="Screenshot 2025-05-01 at 12 26 13" src="https://github.com/user-attachments/assets/b8fe5997-324d-41cb-9882-b4a5ab05bd26" />

https://github.com/user-attachments/assets/fe05b83d-d00c-42ed-8994-c97b0c4de924


In this PR:

https://github.com/user-attachments/assets/0d160690-108f-4ac7-bd71-f39c5fe5b4d9

kapral18 commented:
> Hi Karen, nice enhancement, looking great overall, thank you. I found the behaviour of topN in the `Detection alerts` was changed in this PR. The data source included more than just alert's index in this PR. This rendered the same chart between `Raw events` and `Detection alerts`. Please double check.
> 
> Previous behaviour: <img alt="Screenshot 2025-05-01 at 12 26 13" width="2000" src="https://private-user-images.githubusercontent.com/6295984/439585527-b8fe5997-324d-41cb-9882-b4a5ab05bd26.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYxMTQ5MDAsIm5iZiI6MTc0NjExNDYwMCwicGF0aCI6Ii82Mjk1OTg0LzQzOTU4NTUyNy1iOGZlNTk5Ny0zMjRkLTQxY2ItOTg4Mi1iNGE1YWIwNWJkMjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MDFUMTU1MDAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWQzODlkNmE5NmRlMjNhYTc5ZjQyNDUyOTlkNGM1ZTg4YWY2MDhhM2VlMDIwNDkzNDczYTkwOGEyNGQ3NmQxOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.daiygpOHqqsE-XBqLWUMynL0QRx92RWONvx-Kqp45vQ">
> 
>  Screen.Recording.2025-05-01.at.12.26.25.mov 
> In this PR:
> 
>  Screen.Recording.2025-05-01.at.12.27.38.mov

Hey @angorayc thanks for checking. I can't reproduce your second video. This is what I see in my PR branch locally.

https://github.com/user-attachments/assets/2f89386d-ffe6-4ff4-bdda-6c778c731271

The thing is I haven't touched TopN functionality. I only removed dead props like setQuery and `showLegend` and `setQuery`. So I am not sure how my changes could've affected it, especially when I can't reproduce it.

Can you double check if you are indeed on my PR branch when seeing this?

logeekal commented on x-pack/solutions/security/plugins/security_solution/public/detections/components/alerts_kpis/alerts_histogram_panel/index.tsx:
```diff
@@ -203,24 +214,10 @@ export const AlertsHistogramPanel = memo<AlertsHistogramPanelProps>(
 
     const subtitle = showHistogram
-      ? createEmbeddedDataSubtitle(embeddedDataLoaded, embeddedDataAvailable, totalAlerts)
+      ? createEmbeddedDataSubtitle(embeddedDataAvailable, totalAlerts)
```
Previously we were checking if the embeddable data has been loaded or not but now we are not checking the same. Has something changed on embeddable side?

logeekal commented on x-pack/solutions/security/plugins/security_solution/public/detections/components/alerts_kpis/alerts_histogram_panel/index.tsx:
```diff
@@ -203,24 +214,10 @@ export const AlertsHistogramPanel = memo<AlertsHistogramPanelProps>(
 
     const subtitle = showHistogram
-      ? createEmbeddedDataSubtitle(embeddedDataLoaded, embeddedDataAvailable, totalAlerts)
+      ? createEmbeddedDataSubtitle(embeddedDataAvailable, totalAlerts)
```
Previously we were checking if the embeddable data has been loaded or not but now we are not checking the same. Has something changed on embeddable side?

kapral18 replied:
I simplified the logic and merged loaded & available into just available

logeekal reviewed (approved):
Code review only. LGTM 🚀 . Added one small question.

angorayc commented:
> Can you double check if you are indeed on my PR branch when seeing this?

I ingested data with packetbeat, I have the same observation as before.
I found you have a disabled inspect option in topN under Raw events in your video. Would like to know how did you ingest the data

<img width="1239" alt="Screenshot 2025-05-02 at 09 28 13" src="https://github.com/user-attachments/assets/b9abe0e7-2874-4556-b7b6-2431b9e01314" />

kapral18 committed:
Merge branch 'main' into fix/explore/214733-handle-overview-alerts-trends-crash

* main: (111 commits)
  [ResponseOps][Rules] Cases action title length too long (#219226)
  [main] Sync bundled packages with Package Storage (#219839)
  Fix ignored dynamic templates (#219875)
  Enforce dependency review by kibana-security workflow (#219262)
  [Security Solution] [Detections] Removes tech preview text from eql seq suppression ui (#219870)
  [Security Solution] Fix alerts table potentially not applying alert assignees (#219460)
  fix(slo): alert deletion (#219876)
  [AI4DSOC] fix styling to address cutoff when screen is narrow (#219306)
  [Security Solution][Endpoint] Response action create and history log API updates in of space awareness (#218674)
  Update publish_oas_docs.sh to deploy Kibana Serverless API docs (#219867)
  feat(slo): lock resource installation (#219747)
  [AI4DSOC] Alert flyout code cleanup (#219810)
  [fleet] fixing `isAgentlessDefault` config usage and readability improvements to `isAgentlessSetupDefault` (#219423)
  feat(slo): Bulk delete UI (#219634)
  m1 demo prep (#219588)
  [Security Solution] Replace sourcerer in EQL tab with dataview picker (#218897)
  [AI4DSOC] Attack discovery widget follow up follow up (#219849)
  [AI Assistant] Fix some OpenAI models not accepting temperature for Inference service (#218887)
  Update dependency msw to ~2.7.5 (main) (#219289)
  Use new client URLs in doc link service (#219600)
  ...

angorayc commented:
I think I found the cause: I have `newDataViewPickerEnabled` feature flag enabled.

rylnd reviewed (approved):
API discussion aside, the Detection Engine changes are logically straightforward and low-risk. Approved!

kapral18 committed:
chore: address comments

- remove setTimeout from VisualizationEmbeddableComponent as it was
  causing a race condition and sometimes disabling inspect button
- add accompanying mock for useVisualizationResponse to simplify mocking

kapral18 committed:
Merge branch 'main' into fix/explore/214733-handle-overview-alerts-trends-crash

kapral18 commented:
> Hi Karen, nice enhancement, looking great overall, thank you. I found the behaviour of topN in the `Detection alerts` was changed in this PR. The data source included more than just alert's index in this PR. This rendered the same chart between `Raw events` and `Detection alerts`. Please double check.
> 
> Previous behaviour: <img alt="Screenshot 2025-05-01 at 12 26 13" width="2000" src="https://private-user-images.githubusercontent.com/6295984/439585527-b8fe5997-324d-41cb-9882-b4a5ab05bd26.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYzODMzNzYsIm5iZiI6MTc0NjM4MzA3NiwicGF0aCI6Ii82Mjk1OTg0LzQzOTU4NTUyNy1iOGZlNTk5Ny0zMjRkLTQxY2ItOTg4Mi1iNGE1YWIwNWJkMjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MDRUMTgyNDM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmU5NTU1ZWM3MmFkNTMwNzA4MzljN2U5N2FmODYxZjRjNTUxNWY4YjU0MTgxMzIwZTc3ZTk2NWFiYmY1MGI1OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.2_P5tgT7ThuAPf6sHYQ2lZFapgUBXV_h5uAazhFDw5A">
> 
>  Screen.Recording.2025-05-01.at.12.26.25.mov 
> In this PR:
> 
>  Screen.Recording.2025-05-01.at.12.27.38.mov

As discussed in slack:

- the index pattern is not an issue from this PR. It also happens outside of this PR if `newDataViewPickerEnabled` is enabled.
- I also found another issue with sometimes disabled inspect button. Turns out it was caused by a preexisting setTimeout that was causing a race condition. I pushed the fix.

Thanks for checking.

kapral18 committed:
chore: fix ci regressions

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/298432)
* Commit: 3b4bc228f683f778099998580fee277a3107e78b

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/298432#01969cda-4437-438b-bc58-48c0a077eb97)

### History
* :broken_heart: [Build #298427](https://buildkite.com/elastic/kibana-pull-request/builds/298427) failed d4c41eec40a12828b3dcba6ffaf2ea047bc73906
* :broken_heart: [Build #297385](https://buildkite.com/elastic/kibana-pull-request/builds/297385) failed 4144d8247d39cce3ae25c62be5833b4ace094d73
* :broken_heart: [Build #297083](https://buildkite.com/elastic/kibana-pull-request/builds/297083) failed fd0bbdf258ba12646497ba3e3c6dfac0ca9d7855
* :broken_heart: [Build #297056](https://buildkite.com/elastic/kibana-pull-request/builds/297056) failed a86167d2ef99ae6868b0cb36d0e240432ab7e9a7
* :broken_heart: [Build #297036](https://buildkite.com/elastic/kibana-pull-request/builds/297036) failed 5cb3ee574764d4217406805b812e48aac6ec6dba

cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298432","number":298432,"commit":"3b4bc228f683f778099998580fee277a3107e78b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298427","number":298427,"commit":"d4c41eec40a12828b3dcba6ffaf2ea047bc73906"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297385","number":297385,"commit":"4144d8247d39cce3ae25c62be5833b4ace094d73"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297083","number":297083,"commit":"fd0bbdf258ba12646497ba3e3c6dfac0ca9d7855"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297056","number":297056,"commit":"a86167d2ef99ae6868b0cb36d0e240432ab7e9a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297036","number":297036,"commit":"5cb3ee574764d4217406805b812e48aac6ec6dba"}],"number":298432}
buildkite-pr-comment-->

kapral18 committed:
chore: fix ci regressions part 2

kibanamachine commented:
Starting backport for target branches: 8.17, 8.18, 8.19, 9.0

https://github.com/elastic/kibana/actions/runs/14833434484

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/298483)
* Commit: 853bae7e07ae41731f49523cc8ae774485c10bbe

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/10692211bc7fb078d05ad0499d4cf333fbbfc4b8) | [after](https://github.com/elastic/kibana/commit/853bae7e07ae41731f49523cc8ae774485c10bbe) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.1MB | 9.1MB | -510.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/10692211bc7fb078d05ad0499d4cf333fbbfc4b8) | [after](https://github.com/elastic/kibana/commit/853bae7e07ae41731f49523cc8ae774485c10bbe) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 88.0KB | 88.0KB | +1.0B |



### History
* :broken_heart: [Build #298432](https://buildkite.com/elastic/kibana-pull-request/builds/298432) failed 3b4bc228f683f778099998580fee277a3107e78b
* :broken_heart: [Build #298427](https://buildkite.com/elastic/kibana-pull-request/builds/298427) failed d4c41eec40a12828b3dcba6ffaf2ea047bc73906
* :broken_heart: [Build #297385](https://buildkite.com/elastic/kibana-pull-request/builds/297385) failed 4144d8247d39cce3ae25c62be5833b4ace094d73
* :broken_heart: [Build #297083](https://buildkite.com/elastic/kibana-pull-request/builds/297083) failed fd0bbdf258ba12646497ba3e3c6dfac0ca9d7855
* :broken_heart: [Build #297056](https://buildkite.com/elastic/kibana-pull-request/builds/297056) failed a86167d2ef99ae6868b0cb36d0e240432ab7e9a7

cc @kapral18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298483","number":298483,"commit":"853bae7e07ae41731f49523cc8ae774485c10bbe"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298432","number":298432,"commit":"3b4bc228f683f778099998580fee277a3107e78b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298427","number":298427,"commit":"d4c41eec40a12828b3dcba6ffaf2ea047bc73906"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297385","number":297385,"commit":"4144d8247d39cce3ae25c62be5833b4ace094d73"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297083","number":297083,"commit":"fd0bbdf258ba12646497ba3e3c6dfac0ca9d7855"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/297056","number":297056,"commit":"a86167d2ef99ae6868b0cb36d0e240432ab7e9a7"}],"number":298483}
buildkite-pr-comment-->

kapral18 commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.18|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/220120">](https://github.com/elastic/kibana/pull/220120)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
